### DISCUSSION TITLE
Usage with `zod` and `next-safe-action`

### DISCUSSION THREAD

**Discussion by ScreamZ** [amannn/next-intl#1134] [Q&A]
Hello, I would like to use https://next-safe-action.dev/docs/recipes/i18n as you can see you're injecting the translate function.

Is there any way to get the proper type of this function (on server and client ? getTranslations and useTranslations ?) for something isomorphic ?

I would like to reuse the schema on client and server side.

**Comment by amannn** [comment]
Getting the types for `t` right is unfortunately a bit cumbersome, therefore I generally suggest to not pass the function anywhere but to use it right where it is retrieved.

The linked example code looks like this:

```tsx
async function getSchema() {
  // This is an example of a i18n setup.
  const t = await getTranslations();
  return mySchema(t); // this is the schema that will be used to validate and parse the input
}
```

What is the implementation of `mySchema` here? Can you construct the schema in `getSchema` directly and call `t` as necessary there?

/cc @TheEdoRan

**Reply by ScreamZ** [reply]
Schema is a zod schema most of the time, and you need to pass error translations.

I could rely on passing each key individually, but this requires a lot of code. That's why it could be nice to directly pass the t function.

The use case requires to be working on both the server and the client for me as I'm also using the schema for https://next-safe-action.dev/docs/integrations/react-hook-form

I've tried hacky things like `ReturnType<typeof useTranslation>` but this is hacky and not really efficient, as the signature not really the same as getTranslations

**Reply by amannn** [reply]
It seems like `zod` allows to provide an [`errorMap`](https://zod.dev/ERROR_HANDLING?id=contextual-error-map) to `parse` and `safeParse` that can be used to turn structured errors into localized strings.

Here's the gist with plain zod (without `next-safe-action`):

```tsx
const t = await getTranslations('ZodFormExample');

const result = FormSchema.safeParse(values, {
  errorMap(issue, ctx) {
    let message;
    if (isEqual(issue.path, ['task'])) {
      message = t('taskEmpty');
    }
    if (!message) {
      message = ctx.defaultError;
    }
    return {message};
  }
});
```

([complete example](https://github.com/amannn/next-intl/blob/main/examples/example-app-router-playground/src/app/%5Blocale%5D/actions/ZodFormExample.tsx#L29-L42))

Is the call to `parse` hidden inside `next-safe-action`? If yes, maybe the library could provide an option to forward the `params` argument to the internal call to `zod`?

In this example, the schema can be constructed separately from the call that uses it to parse messages (and might want to return localized messages).

/cc @TheEdoRan

**Reply by TheEdoRan** [reply]
Hi @amannn, I never used i18n solutions with Next.js personally, the [i18n recipe](https://next-safe-action.dev/docs/recipes/i18n) on the website is based on [this comment](https://github.com/TheEdoRan/next-safe-action/discussions/111#discussioncomment-9408999) left by @ali-alharthi in the related discussion, and it's mainly there to show how to pass an awaited schema to the action.

That said, next-safe-action handles parsing and validation internally, using [TypeSchema ](https://typeschema.com/)to support multiple validation libraries, so it doesn't expose the `params`. How do libraries like [zod-i18n](https://github.com/aiji42/zod-i18n) work? That way seems quite clean.

**Reply by amannn** [reply]
Thanks for chiming in here @TheEdoRan!

 `zod-i18n` works by [calling `z.setErrorMap`](https://github.com/aiji42/zod-i18n?tab=readme-ov-file#how-to-use). That certainly works, but that sets the errors globally. Especially for a use case like Server Actions, where a server is expected to be stateless, I guess it would be beneficial to avoid overriding the global error map on every request (could also lead to race conditions when multiple requests are handled concurrently).

I can think of the following options:
1. Providing localized labels while parsing using an existing schema ([my example above](https://github.com/amannn/next-intl/discussions/1134#discussioncomment-9807622))
2. Creating a new schema on every request while attaching localized labels (that's what the sample from @ali-alharthi does: [schema.ts](https://github.com/ali-alharthi/i18n-example/blob/main/src/server/schema.ts)). I think this implementation has a few downsides: a) It's cumbersome to type the `t` function correctly (the original request in this issue), b) the schema needs to be created on every request, c) all localized labels have to be generated upfront, even though they might not be used.
3. Does `next-safe-action` provide a mechanism to catch a validation error and transform the data structure? I've seen the `onError` callback in your docs, but I'm not really familiar with the library. Does the callback consider a return value?

If (1) isn't supported by `next-safe-action` currently, could (3) maybe be a good alternative?

**Reply by TheEdoRan** [reply]
3. Yes, via [`handleValidationErrorsShape`](https://next-safe-action.dev/docs/recipes/customize-validation-errors-format) passed to the `schema` method, this might be the solution.

**Reply by amannn** [reply]
> Yes, via [`handleValidationErrorsShape`](https://next-safe-action.dev/docs/recipes/customize-validation-errors-format) passed to the `schema` method, this might be the solution.

That could be something for you to have a look at @ScreamZ.

**Reply by ali-alharthi** [reply]
@amannn Regarding point no. 2, I do agree on the downsides you have mentioned.
However, the provided example was to try and explain what @jeltehomminga was talking about in his discussion as we faced the same challenge as well in one of our projects. Also, it seemed the "easier way" for us to get `next-safe-action`, `zod` and `next-intl` to work at that time.

**Reply by ScreamZ** [reply]
Hello boyz!

Just tested the given solution from @amannn using the new wrapper for hook form by @TheEdoRan 

```ts
  const { form, action } = useHookFormAction(
    sendEmailAuthAction,
    zodResolver(sendEmailAuthSchema, {
      errorMap: (issues) => {
        if (issues.path.includes("email")) {
          return { message: t("Errors.invalid-email") };
        } else {
          return { message: t("Errors.generic") };
        }
      },
    })
  );
``` 

Here is the code. This works, but this remains very very verbose for not much. Also you have to write deep code and not just rely on path to know at which segment the error occurs (is it in .email ? .refine ? other)

Well I still recommended injecting the t function as simple as follow:

```ts
// i18n/types.ts
import type { useTranslations } from "next-intl";
import type { getTranslations } from "next-intl/server";

export type TFn =
  | ReturnType<typeof useTranslations<never>>
  | Awaited<ReturnType<typeof getTranslations<never>>>;
``` 

Then usage like this

```ts
// In a separated file (not the one with "use server"

// Import the previous TFn
export const sendEmailAuthSchema = (t: TFn) =>
  z.object({
    email: z
      .string()
      .email(t("Errors.invalid-email"))
      .refine((email) => !email.includes("+"), t("Errors.no-subdomain")),
  });
``` 

```ts
// Client side
const t = useTranslations()
const { form, action } = useHookFormAction(sendEmailAuthAction(t), zodResolver(sendEmailAuthSchema(t)));
``` 

```ts
 // Server side
export const sendEmailAuthAction = publicAction
  .schema(async () => sendEmailAuthSchema(await getTranslations()))
  .action(async ({ parsedInput: { email } }) => {
// ...
})
``` 

Easy peasy, and cross platform support.

**Comment by ScreamZ** [comment]
I'm thinking of using something like https://nextjs.org/docs/pages/building-your-application/optimizing/lazy-loading#with-external-libraries to conditionnaly use getTranslations or useTranslations depending on we are on server or client side. By checking `typeof window`
